### ISSUE TITLE
Stand-alone and AmberTools parmed

### ISSUE THREAD

**Issue by jchodera** [ParmEd/ParmEd#1015]
Installing AmberTools seems to wipe out the installation of stand-alone parmed, and vice-versa. 

@swails : Do you have a suggested resolution to this? Are there plans to remove parmed from the conda-installable AmberTools, and have it just require the parmed package as a dependency? Or vice-versa?

**Comment by hainm** [comment]
You can do `pip uninstall parmed` first (works for AmberTools18 conda version).

> Are there plans to remove parmed from the conda-installable AmberTools, and have it just require the parmed package as a dependency? Or vice-versa?

I think this's a complicated process for AT distribution.

**Comment by jchodera** [comment]
If a tool uses both `ambertools` and `parmed` conda recipes as dependencies, the install order is not guaranteed, and which parmed version you end up with depends on the order in which they were installed.

We can't tell conda dependencies to run `pip uninstall parmed` first, unless we create a pre-install script in the conda recipe for parmed that tries to uninstall `parmed` if it already exists---but even then, installing `ambertools` would overwrite it.

There's no sane way I can figure to have parmed belong *both* to the `ambertools` conda distribution *and* to its own `conda` distribution. The only way to keep it in `ambertools` would be to release an `ambertools` update each time a new release of `parmed` is cut.

**Comment by jchodera** [comment]
Hm, if we add `ambertools` as a requirement for `parmed` distributed via omnia, we can at least require that `parmed` is installed *after* `ambertools` the first time, but I don't think this would re-install `parmed` if `ambertools` is later upgrade.

One possibility is pin the `ambertools` dependency version for each `parmed` conda release on omnia?

**Comment by swails** [comment]
It seems to me to make perfect sense to have AmberTools depend on (and *not* bundle) ParmEd.  Especially since ParmEd gets built unversioned within AmberTools.

**Comment by swails** [comment]
> I think this's a complicated process for AT distribution.

I don't think it's *that* complicated.  We just need to have the conda recipe *not* build ParmEd.  But you (@hainm) did all the AmberTools conda stuff -- would you have a chance to take a quick look at this?

**Comment by hainm** [comment]
> It seems to me to make perfect sense to have AmberTools depend on (and not bundle) ParmEd. 

Then it's not AmberTools anymore. The binary distribution should closely match to the version compiled from source code.  

cc @ dacase about this.

**Comment by hainm** [comment]
> I don't think it's that complicated. We just need to have the conda recipe not build ParmEd. But you (@hainm) did all the AmberTools conda stuff -- would you have a chance to take a quick look at this?

Yeah, we could do it. Just delete the ParmEd folder after compiling ambertools. We need, then, to build parmed and push it to RCSB server (I manually copy (from several ssh steps) AmberTools binary to the server due to firewall issue). I don't have a clear mind about this since I have little time for AMBER now and I am not sure who's eager to actively maintain it.  

But the question is: What's the chance that user requires a parmed version that is not bundled in AT, given that ParmEd is quite stable and not super actively developed anymore.

**Comment by hainm** [comment]
> Especially since ParmEd gets built unversioned within AmberTools.

Can we make it versioned? (but not sure if that helps the case here).

**Comment by swails** [comment]
> What's the chance that user requires a parmed version that is not bundled in AT, given that ParmEd is quite stable and not super actively developed anymore.

It's not that unusual.  There have been a few bugs fixed in ParmEd since AmberTools was released, at least one of which a user of AmberTools recently ran into on the mailing list.

> Then it's not AmberTools anymore. The binary distribution should closely match to the version compiled from source code.

You can't have it both ways, though :-P.  It can't be both unnecessary because the project is not developed that quickly anymore *and* a pollution that makes it no longer AmberTools because it's a few commits ahead.  And besides, AmberTools can simply depend on the specific version of ParmEd that was cut when AmberTools was released, right?  We ought to make it easy to upgrade *just* ParmEd when AmberTools is installed via conda.  [Practicality beats purity](https://www.python.org/dev/peps/pep-0020/#id3) and all that.

**Comment by hainm** [comment]
> It's not that unusual. There have been a few bugs fixed in ParmEd since AmberTools was released, at least one of which a user of AmberTools recently ran into on the mailing list.

So user can install the newer version of Parmed and the one in AT will be overwritten. I don't see any drawback here. :D 

> And besides, AmberTools can simply depend on the specific version of ParmEd that was cut when AmberTools was released, right? 

If AmberTools depends on ParmEd 2.5 (just saying), and users install ParmEd 2.6, I don't think conda will allow that due to package conflicting (someone needs to confirm this).

> Practicality beats purity and all that.

Yeah, I agree. :).

**Comment by hainm** [comment]
And someone also argues about the overwriting of standalone cpptraj, reduce, pytraj ... programs too. :D

**Comment by hainm** [comment]
Ah, just FYI in case you have not noticed: I updated AT18 a while ago so most of Python programs (pytraj, parmed, pdb4amber) use setuptools for distribution, this will allow user to upgrade to newer versions; (otherwise there would be message saying packages distributed by distutils could not be deleted completely).

**Comment by jchodera** [comment]
> If AmberTools depends on ParmEd 2.5 (just saying), and users install ParmEd 2.6, I don't think conda will allow that due to package conflicting (someone needs to confirm this).

AmberTools can depend on `parmed >=2.5` and it will make sure that users will have at least ParmEd 2.5 installed. Users can still upgrade ParmEd, and provided ParmEd is backwards-compatible, everything will work.

**Comment by jchodera** [comment]
> So user can install the newer version of Parmed and the one in AT will be overwritten. I don't see any drawback here. :D

The issue is that conda doesn't know anything about how to do this correctly. If our tools depend on a recent version of ParmEd and AmberTools 18, we don't have any way to ensure this in the conda recipe except to ask the user execute a manual sequence of steps to upgrade ParmEd seprarately after AmberTools has been installed in order to get the software to work. We're trying to make it easier to distribute software to maximize usability, not harder...

**Comment by dacase** [comment]
On Fri, Nov 02, 2018, someone wrote:

> > It seems to me to make perfect sense to have AmberTools depend on (and
> > not bundle) ParmEd.

My concerns here:

a. AmberTools users should have ready access to the full source code of
all components.  We would have to carefully document how this could be
accomplished if parmed were no longer bundled.

b. The reason for issue (a) is to encourage uses to study and modify the
code.  We aleady make this somewhat difficult by using setuptools: if a
user wants to try out any changes to parmed or pdb4amber (even to just
print out some extra information to follow how the code works), they can't
just edit the python code in AMBERHOME/src/{parmed,pdb4amber}, but have to
re-run "make install" after each such change. That is, even in the current
mode, the .py files effectively act like compiled code, since they have to
be re-packaged into "egg" files after each change.  But I don't think we
actually explain this anywhere.

If users were to obtain parmed via a conda install command, we have to
make sure they understand how to make test modifications "active" within
AmberTools.

c. I generally want to promote the idea of making things easy for users,
even at the expense of adding some extra work for developers.  This is
of course is a balancing act.  But generally, the more dependencies
something has, the more complex the installation and operation
instructions become, to the detriment of ease of use.

d. Above doesn't mean we shouldn't make this change, just that we
carefully consider its implications.

...dac

**Comment by swails** [comment]
@dacase - you misunderstand my proposal. My proposal is to remove bundling ParmEd from the AmberTools conda install only. I’m not proposing any changes to our “traditional” release model. The source code will still be bundled and installed as a default part of the AmberTools build process. 

Users should see almost no change, since AmberTools conda will still pull in ParmEd (so it’s effectively bundled), but there will be a separation that allows independent upgrades of ParmEd in a conda environment only.

**Comment by jchodera** [comment]
@dacase: Just to add to @swails comment: Adding `parmed` as a dependency to the `ambertools` conda package would still deliver users the *exact* same functionality as before, but with the added bonus that they can also later upgrade `parmed` independently via conda without breaking everything. Other routes for delivering AmberTools not via conda should remain unchanged.

**Comment by dacase** [comment]
This may be a bit off-topic, but applies to how ambertools gets distributed to users, especially via conda:

One of the nice things about python is that it is interpreted, so that users can (in simple cases) edit the files and play around with ideas, perhaps as a part of debugging/understanding, or perhaps as a part of playing with feature extensions.  The user doesn't have to know how to re-build the package after an edit--it just "happens" (`.pyc` files get automagically updated, etc.)

Even with compiled modules, there is generally a lot of plain python code around, and we want to encourage people to explore/debug/extend it.

If we just tell users to `conda install ambertools` (whether or not that bundles parmed or installs it as a prerequisite) we have hidden the location of the active code (deep inside some egg directory or zip file.)  I think we should have a section in the documentation that explains how to find (and modify) the .py files that are actually being interpreted.  

And, to consider the tradeoffs involved in using setuptools and their kin: are there times when the conveniences of this sort of packaging and versioning outweighed by the added complexities?  Are there times when simply having a directory containing human readable .py files (including __init__.py)
is the better way to go?

**Comment by swails** [comment]
I think it would be interesting to know who uses conda to install AmberTools and, more importantly, _why_ they do that.

If (as I suspect) the conda package's primary use-case is for situations like this where it can be deployed as a black-box dependency (in some kind of automated environment like a docker container or CI/CD pipeline), there's likely no demand for this kind of transparency (and an experienced Python user will have no trouble doing

```python
import parmed
print(parmed.__file__)
```

to locate the files if they want to edit it as you describe).

If conda is being used in place of the "traditional" (source-based) installation, then we probably need a better process to ensure that updates result in an updated conda package being pushed to conda-forge (or wherever it's going).

I know we can get stats on conda downloads for AmberTools -- does anybody with visibility into that know what those statistics are?  Because I think we ought to optimize the package for what it's actual use is in the field.

Documentation is not a bad idea, but my personal advice to anybody that wants to tinker is to download the source distribution to do that (or the standalone Python tools if they want to tinker with just ParmEd).

Open to thoughts.

Thanks!

**Comment by jchodera** [comment]
> Documentation is not a bad idea, but my personal advice to anybody that wants to tinker is to download the source distribution to do that (or the standalone Python tools if they want to tinker with just ParmEd).

This is *definitely* the practice we want to encourage, since it's good industry standard best practice. We *don't* want users just editing their installed Python files in their miniconda distribution, since that can cause all sorts of trouble and nobody recommends doing that.

If part of our objective here is to encourage good behavior that will be useful to users whether they get industry or academic jobs, we want to recommend best practices whenever possible, even if it's marginally more work to `git clone` the current repo and `python setup.py develop` to install than just editing installed files in place.

**Comment by hainm** [comment]
> The issue is that conda doesn't know anything about how to do this correctly. If our tools depend on a recent version of ParmEd and AmberTools 18, we don't have any way to ensure this in the conda recipe except to ask the user execute a manual sequence of steps to upgrade ParmEd seprarately after AmberTools has been installed in order to get the software to work. We're trying to make it easier to distribute software to maximize usability, not harder...

How's about writing a script to modify the original AmberTools:
- Delete parmed
- Update meta data to make AT depend on ParmEd.

Then you can distribute that modified version in omnia? 

PS: I already wrote a context manager to edit the conda package: https://github.com/Amber-MD/ambertools-binary-build/blob/master/conda_tools/edit_package.py#L44

**Comment by jchodera** [comment]
@hainm: I like the idea! Would be happy to do this.

Would you be able to share the recipe you use to build the AmberTools conda package?

**Comment by jchodera** [comment]
Pinging @hainm: To remove parmed, are you suggesting I *edit* the conda package using [your tool](https://github.com/Amber-MD/ambertools-binary-build/blob/master/conda_tools/edit_package.py#L44) or rebuild a new conda package with a modified conda recipe?

**Comment by hainm** [comment]
@jchodera  yeah, I suggest to edit existing AT file since AT conda build is not a typical build with a given conda recipe. (We pack all python builds (py2.4, 3.5, ...) into a single tarfile).

PS: I normally build all AT (conda build + non-conda build) with my macbook for both macos and linux (using docker) with a single build script :D 

https://github.com/Amber-MD/ambertools-binary-build/blob/master/build_all.py
In short: kind of complicated build. :D

**Comment by jchodera** [comment]
@hainm: Does the current tarfile also support Python 3.7, now that conda-forge has moved to supporting only python 2.7, 3.6, and 3.7?

**Comment by jchodera** [comment]
@hainm: I also wonder if we'd be better off with an `ambermini` version as well. Do the build scripts work with the `--task ambermini` option? I notice that some files appear to be missing, such as `conda_tools/load_functions.sh`.

**Comment by jchodera** [comment]
@hainm: Last question for now---I promise! Is there a current infrastructure maintainer for the AmberTools conda packages besides yourself, @hainm? I hate to bug you at your current job if there is someone else we should be talking to! And if not, maybe we could help with AmberTools conda package release infrastructure, potentially along with @lnaden and the MolSSI?

**Comment by hainm** [comment]
> Does the current AmberTools conda package tarfile also support Python 3.7, now that conda-forge has moved to supporting only python 2.7, 3.6, and 3.7?

No, AmberTools 18 only supports py 2.7, 3.4, 3.5, 3.6. There was no 3.7 yet when I built the binary. I can  update py 3.7 when I come back to US in next few weeks. 

> Do the build scripts work with the --task ambermini option?

Not anymore. I think it's quite easy to just install the full AmberTools to avoid maintenance burden. But if omnia wants to maintain ambermini, I think it's quite easy to make a receipt for it.

> Last question for now---I promise!

No problem, I love to answer questions :P.

>  Is there a current infrastructure maintainer for the AmberTools conda packages besides yourself, @hainm?

No, only me. I think most of Amber developers are volunteer so people only work on things they're interested. 

> And if not, maybe we could help with AmberTools conda package release infrastructure, potentially along with @Lnaden and the MolSSI?

That would be great. Currently we can not use travis due to the very long build time for AT with multiple python versions (that will be packed in a single package).

**Comment by jchodera** [comment]
@dacase: That sounds exactly like what we need! We'd definitely be interested in helping craft (and ideally find a way to automatically maintain up to date) a version of this on `omnia`.

We could re-use `ambermini` (since this would now be version 18, the last `ambermini` release was 16, and we would presumably include at least the same minimal set of tools), or go with something like `amberlite` or `ambermicro`.

It looks like I don't have an active account on the gitlab site. Would you be willing to send an invitation to `john.chodera@choderalab.org`?

**Comment by dacase** [comment]
Just an update: the recent `amber_omnia` package (`conda install -c ambermd amber_omna`) doesn't include ParmEd, anticipating that the toolset will get that directly from the github `ParmEd` site.  So the initial impetus for this long thread should be mooted.  Also, I think(?) this `amber_omnia` package is independent of python (and hence python version).  But this is still experimental, and I'm very open to improving it.
